### PR TITLE
Adding coverage for putUnlessExists, with fix for NPE

### PR DESCRIPTION
**Pull request by gokcanozakdag**
**Goals (and why)**:
- Adding test coverage for the case when we try to commit a transaction that is already committed; possibly as a result of putUnlessExists retry logic. Also fixing NPE while logging the `tableRef`.

**Implementation Description (bullets)**:
- Forcing timestampService to return same timestamp (`commitTs`) when fresh timestamp is requested. Putting a transaction with `startTs` and `commitTs` to `transactionService` before calling `commit` on transaction, which will cause `putUnlessExists` to throw `KeyAlreadyExists`.

**Concerns (what feedback would you like?)**:
- Is there a better way to test this case?
- Checking `tableRef != null` and logging accordingly seems a bit ugly to me, not sure if logging without tableRef is useful at all. If not, we can just remove that part, and log only if tableRef is not null.

**Where should we start reviewing?**:
- SnapshotTransaction.java

**Priority (whenever / two weeks / yesterday)**:
- Not high priority, P2.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/3205)

### CODE REVIEW

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 1]
```diff
@@ -1014,6 +1015,35 @@ public void cleanup() {}
         assertEquals(ImmutableList.of(firstCell, secondCell), cells);
     }
 
+    @Test
+    public void commitDoesNotThrowIfAlreadySuccessfullyCommitted() {
```
Comment (gsheasby) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: what should happen if we try to commit, but somebody has rolled us back - is that case covered?

**Comment by gsheasby** [comment]
Next action: @gokcanozakdag to add the above mentioned test case

**Comment by gokcanozakdag** [comment]
@gsheasby @sandorw I have revised the pr, can you take a final look?

File: docs/source/release_notes/release-notes.rst [hunk 2]
```diff
@@ -66,6 +66,11 @@ develop
            transactions had significantly impaired performance if a database node was down.
            (`Pull Request <https://github.com/palantir/atlasdb/pull/3199>`__)
 
+    *    - |fixed|
+         - Fixed an issue occurring after a failure to putUnlessExists a commit timestamp that was causing an NPE leading to a confusing error message.
```
Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: The release note should capture in which context we would run into this issue in the field.

**Comment by gsheasby** [comment]
The new test failed :-(
```
[00:04:30] com.palantir.atlasdb.transaction.impl.SnapshotTransactionTest > commitThrowsIfRolledBackAtCommitTime STARTED
[00:04:30] 
[00:04:30] com.palantir.atlasdb.transaction.impl.SnapshotTransactionTest > commitThrowsIfRolledBackAtCommitTime FAILED
[00:04:30]     java.lang.AssertionError: 
[00:04:30]     Expecting:
[00:04:30]      <java.lang.AssertionError: BUG: Someone tried to roll back our transaction but our locks were still valid; this is not allowed. Held immutable timestamp lock: Optional.empty; held commit locks: LockToken{requestId=38450ec9-0151-bdc8-3e52-74c490b057f6}>
[00:04:30]     to be an instance of:
[00:04:30]      <com.palantir.atlasdb.transaction.api.TransactionCommitFailedException>
[00:04:30]     but was instance of:
[00:04:30]      <java.lang.AssertionError>
[00:04:30]         at com.palantir.atlasdb.transaction.impl.SnapshotTransactionTest.commitThrowsIfRolledBackAtCommitTime(SnapshotTransactionTest.java:1057)
[00:04:30] 
```

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 3]
```diff
@@ -1048,15 +1050,50 @@ public void commitDoesNotThrowIfAlreadySuccessfullyCommitted() {
...
+        transactionService.putUnlessExists(snapshot.getTimestamp(), TransactionConstants.FAILED_COMMIT_TS);
 
-        tx.put(TABLE, ImmutableMap.of(cell, PtBytes.toBytes("value")));
-        assertThatThrownBy(() -> tx.commit()).isInstanceOf(TransactionCommitFailedException.class);
+        snapshot.put(TABLE, ImmutableMap.of(cell, PtBytes.toBytes("value")));
+        assertThatThrownBy(() -> snapshot.commit()).isInstanceOf(TransactionLockTimeoutException.class);
     }
 
+
+
```
Comment (gsheasby) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: nit: triple-blank line

**Pull request merged**